src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -9985,30 +10037,39 @@ protected override void VisitCatchBlock(BoundCatchBlock node, ref LocalState fin
 
         public override BoundNode? VisitDeconstructValuePlaceholder(BoundDeconstructValuePlaceholder node)
         {
+            // These placeholders don't yet follow proper placeholder discipline
+            Debug.Assert(!PlaceholderMustBeRegistered(node));
             SetNotNullResult(node);
             return null;
         }
 
         public override BoundNode? VisitObjectOrCollectionValuePlaceholder(BoundObjectOrCollectionValuePlaceholder node)
         {
+            // These placeholders don't yet follow proper placeholder discipline
+            Debug.Assert(!PlaceholderMustBeRegistered(node));
             SetNotNullResult(node);
             return null;
         }
 
-        [MemberNotNull(nameof(_resultForPlaceholdersOpt))]
-        private void EnsurePlaceholdersToResultMap()
-        {
-            _resultForPlaceholdersOpt ??= PooledDictionary<BoundValuePlaceholderBase, (BoundExpression Replacement, VisitResult Result)>.GetInstance();
-        }
-
         public override BoundNode? VisitAwaitableValuePlaceholder(BoundAwaitableValuePlaceholder node)
         {
+            // These placeholders don't always follow proper placeholder discipline yet
+            Debug.Assert(!PlaceholderMustBeRegistered(node));
             VisitPlaceholderWithReplacement(node);
             return null;
         }
 
         private void VisitPlaceholderWithReplacement(BoundValuePlaceholderBase node)
         {
+            if (PlaceholderMustBeRegistered(node))
```

Comment:
>PlaceholderMustBeRegistered [](http://example.com/codeflow?start=16&length=27)

I am not sure if it is worth crashing compiler if the requirement is violated and duplicating the logic to achieve that. It feels like a simple assert within the original `else` branch would be sufficient.